### PR TITLE
New Framework for Anomaly Detection Function

### PR DESCRIPTION
**Pull request by cyenjung**
The new framework breaks down the flow of an anomaly function into multiple reusable and pluggable components. In the new framework, an anomaly function consists the following models: 1. data model, 2. transformation functions, 3. prediction model, 4. detection model, and 5. merge model.

Overview of these models:
1. Data Model defines the time ranges of the training data for the current time series.
2. Transformation defines the transformation to be done before anomaly detection such as smoothing.
3. Prediction model takes as input the training data and output a trained prediction model. 
4. Detection model takes as input the trained prediction model and tests if there exist any anomalies on the current time series.
5. Merge model takes as input a merged anomaly and compute its weight and score.

Current implemented components for each model:
1. Data Model:
  a. SeasonalDataModel, which returns multiple non-consecutive time ranges of baselines of the observed time series.
  b. NoopDataModel, which returns an empty list of time ranges.
2. Transformation Function:
  a. MovingAverageSmoothingFunction, which smooth the given time series using moving average algorithm.
3. PredicationModel:
  a. SeasonalAveragePredictionModel, which returns the average values of multiple baselines.
  b. NoopPredictionModel, which does no-op during prediction.
4. DetectionModel:
  a. SimpleThresholdDetectionModel, which compare the observed and expected time series and returns the data point where the difference between observed and expected values exceeds user-specified threshold.
  b. NoopModel, which does no-op during testing.
5. MergeModel:
  a. AverageAnomalyMergeModel, which average the weight and score of the raw anomalies that are contained by the merged anomaly.
  b. SimplePercentageMergeModel, which uses the average percentage change between observed and expected time series as weight and average the scores of all raw anomalies of the merge anomaly.
  c. NoopMergeModel, which does no-op.

During anomaly detection, the input is stored in an AnomalyDetectionContext and important intermediate results and final detection results are appended to the context.

Moreover, AnomalyDetectionExecutor performs the algorithmic flow of anomaly detection, which applies transformation functions on all time series, trains prediction model, and detects anomalies. 

An enhanced version of WeekOverWeekRule is implemented, which provides smoothing and can use multiple weeks of baselines as the expected time series.

Test Has Done:
An unit test is included. The test applies moving average smoothing and use the average of 2 weeks baseline as the expected time series with the WeekOverWeekRule.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/AbstractAnomalyDetectionFunction.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+public abstract class AbstractAnomalyDetectionFunction implements AnomalyDetectionFunction {
+  protected final Logger LOGGER = LoggerFactory.getLogger(this.getClass());
+
+  protected AnomalyFunctionDTO spec;
+  protected Properties properties;
+
+  protected DataModel dataModel = new NoopDataModel();
+  protected List<TransformationFunction> currentTimeSeriesTransformationChain = new ArrayList<>();
+  protected List<TransformationFunction> baselineTimeSeriesTransformationChain = new ArrayList<>();
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/AbstractAnomalyDetectionFunction.java]: please take a look at "chain of responsibility" design pattern and how we can trigger "apply" method one after another.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/WeekOverWeekRule.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    detectionModel.init(this.properties);
+  }
+
+  private void initPropertiesForDataModel(String baselineProp) {
+    // The basic settings for w/w
+    this.properties.setProperty(SeasonalDataModel.SEASONAL_PERIOD, "1");
+    this.properties.setProperty(SeasonalDataModel.SEASONAL_SIZE, "7");
+    this.properties.setProperty(SeasonalDataModel.SEASONAL_UNIT, "DAYS");
+    // Change the setting for different w/w types
+    if ("w/2w".equals(baselineProp)) {
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/WeekOverWeekRule.java]: Keep Compare Mode consistent with rest of the code... eg w/2w vs Wo2W ?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/context/TimeSeriesKey.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+  public void setDimensionMap(DimensionMap dimensionMap) {
+    this.dimensionMap = dimensionMap;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (o instanceof TimeSeriesKey) {
+      TimeSeriesKey other = (TimeSeriesKey) o;
+      return ObjectUtils.equals(this, other);
```
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/context/TimeSeriesKey.java]: +1
I think there's a good implementation in the bean objects

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/WeekOverWeekRule.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    } else if ("w/3w".equals(baselineProp)) {
+      this.properties.setProperty(SeasonalDataModel.SEASONAL_SIZE, "21");
+    } else if ("w/4w".equals(baselineProp)) {
+      this.properties.setProperty(SeasonalDataModel.SEASONAL_SIZE, "28");
+    } else if ("w/2wAvg".equals(baselineProp)) {
+      this.properties.setProperty(SeasonalDataModel.SEASONAL_PERIOD, "2");
+    } else if ("w/3wAvg".equals(baselineProp)) {
+      this.properties.setProperty(SeasonalDataModel.SEASONAL_PERIOD, "3");
+    } else if ("w/4wAvg".equals(baselineProp)) {
+      this.properties.setProperty(SeasonalDataModel.SEASONAL_PERIOD, "4");
```
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/WeekOverWeekRule.java]: these strings would be better off as an enum in a constants file

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/Utils.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
+package com.linkedin.thirdeye.anomalydetection;
```
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/Utils.java]: We already have a Utils class. You could either use that one itself, or name this to AnomalyUtils to be more specific?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/Utils.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  /**
+   * Returns the bucket size of time series in millis given the keys to retrieve the values from the
+   * properties.
+   *
+   * @param bucketSizeKey the key to retrieve the parameter for bucket size.
+   * @param bucketUnitKey the key to retrieve the parameter for bucket unit.
+   * @param properties the properties that contains the parameters.
+   * @return the bucket size in millis.
+   */
+  public static long getBucketInMillis(String bucketSizeKey, String bucketUnitKey, Properties properties) {
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/Utils.java]: Renamed to AnomalyDetectionUtils.java

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/context/TimeSeriesKey.java (hunk 3 above)
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/context/TimeSeriesKey.java]: Fixed.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/AbstractAnomalyDetectionFunction.java (hunk 1 above)
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/AbstractAnomalyDetectionFunction.java]: Added to TODO list

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/WeekOverWeekRule.java (hunk 2 above)
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/WeekOverWeekRule.java]: The naming is compatible with the current naming in the database, which should be deprecated after migrating to the new framework.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/Utils.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+public class Utils {
+  /**
+   * Parses an anomaly function specification and returns the corresponding property object.
+   *
+   * @param spec the anomaly function spec
+   * @return the corresponding property object
+   * @throws IOException
+   */
+  public static Properties parseSpec(AnomalyFunctionDTO spec) throws IOException {
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/Utils.java]: Done

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/context/AnomalyDetectionContext.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+import com.linkedin.thirdeye.anomalydetection.function.AnomalyDetectionFunction;
+import com.linkedin.thirdeye.anomalydetection.model.prediction.PredictionModel;
+import com.linkedin.thirdeye.datalayer.dto.RawAnomalyResultDTO;
+import java.util.List;
+
+/**
+ * The context for performing an anomaly detection on the sets of time series from the same
+ * dimension and metric.
+ */
+public class AnomalyDetectionContext {
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/context/AnomalyDetectionContext.java]: Done

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/context/TimeSeries.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+/**
+ * Time series for anomaly detection, which contains the pairs of timestamp and data values from
+ * one time series.
+ */
+public class TimeSeries {
+
+  /**
+   * Timestamps to values.
+   */
+  private Map<Long, Double> timeSeries = new HashMap<>();
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/context/TimeSeries.java]: Good idea. Changed to TreeMap.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/WeekOverWeekRule.java (hunk 4 above)
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/WeekOverWeekRule.java]: Agree. I have to use these naming for backward compatibility. I will deprecate this part after migrating to new framework.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/Utils.java (hunk 5 above)
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/Utils.java]: Renamed to AnomalyDetectionUtils.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/WeekOverWeekRule.java (hunk 4 above)
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/WeekOverWeekRule.java]: Added a backward compatible function to parse the baseline property. The function should generic enough without using if-else and switch statement.

**Pull request merged**